components/engine_rocks/src/misc.rs

```diff
@@ -1,90 +1,187 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
 use crate::engine::RocksEngine;
-use crate::util;
+use crate::import::RocksIngestExternalFileOptions;
+use crate::sst::RocksSstWriterBuilder;
+use crate::{util, RocksSstWriter};
 use engine_traits::{
-    CFNamesExt, IterOptions, Iterable, Iterator, MiscExt, Mutable, Range, Result, WriteBatchExt,
-    ALL_CFS, CF_LOCK, MAX_DELETE_BATCH_COUNT,
+    CFHandleExt, CFNamesExt, DeleteStrategy, ImportExt, IngestExternalFileOptions, IterOptions,
+    Iterable, Iterator, MiscExt, Mutable, Range, Result, SstWriter, SstWriterBuilder,
+    WriteBatchExt, ALL_CFS,
 };
 use rocksdb::Range as RocksRange;
 use tikv_util::keybuilder::KeyBuilder;
 
-impl MiscExt for RocksEngine {
-    fn flush(&self, sync: bool) -> Result<()> {
-        Ok(self.as_inner().flush(sync)?)
-    }
+pub const MAX_DELETE_COUNT_BY_KEY: usize = 2048;
 
-    fn flush_cf(&self, cf: &str, sync: bool) -> Result<()> {
-        let handle = util::get_cf_handle(self.as_inner(), cf)?;
-        Ok(self.as_inner().flush_cf(handle, sync)?)
+impl RocksEngine {
+    fn is_titan(&self) -> bool {
+        self.as_inner().is_titan()
     }
 
-    fn delete_files_in_range_cf(
+    // We store all data which would be deleted in memory at first because the data of region will never be larger than
+    // max-region-size.
+    fn delete_all_in_range_cf_by_ingest(
         &self,
         cf: &str,
-        start_key: &[u8],
-        end_key: &[u8],
-        include_end: bool,
+        sst_path: String,
+        ranges: &[Range],
     ) -> Result<()> {
-        let handle = util::get_cf_handle(self.as_inner(), cf)?;
-        Ok(self
-            .as_inner()
-            .delete_files_in_range_cf(handle, start_key, end_key, include_end)?)
-    }
-
-    fn delete_blob_files_in_range_cf(
-        &self,
-        cf: &str,
-        start_key: &[u8],
-        end_key: &[u8],
-        include_end: bool,
-    ) -> Result<()> {
-        if self.as_inner().is_titan() {
-            let handle = util::get_cf_handle(self.as_inner(), cf)?;
-            self.as_inner().delete_blob_files_in_range_cf(
-                handle,
-                start_key,
-                end_key,
-                include_end,
-            )?;
+        let mut ranges = ranges.to_owned();
+        ranges.sort_by(|a, b| a.start_key.cmp(b.end_key));
+        let mut opts = IterOptions::new(None, None, false);
```

Comment:
It's better to add lower_bound and upper_bound for the iterator.

Reply:
But we need to seek multiple range....

Reply:
So use the first range's start_key as lower_bound, use the last range's end_key as upper_bound? And please notice that here `keys` must be start with `z`.